src/recycle/core.clj

```diff
@@ -0,0 +1,184 @@
+(ns recycle.core
+  (:require [clojure.core.async :as a]
+            [recycle.util :as u]))
+
+;; --- Helpers
+
+
+(defn service?
+  [v]
+  (::service v))
+
+(defn service-factory?
+  [v]
+  (::factory v))
+
+;; --- Main API
+
+(def ^:dynamic *timeout*
+  "Maximum number of milliseconds to wait for a response from a
+  service. This variable by default is nil, meaning the value used
+  when creating the service is used."
+  nil)
+
+(defn- default-receive
+  "A default implementation for `:receive`."
+  [& _]
+  (ex-info "service do not implement :receive" {}))
+
+(declare initialize-service)
+(declare start!)
+
+(defn service
+  "Create a service using the spec map as a blueprint.
+  The spec map may contain any of the following keys (all of the are
+  optional):
+
+   - `:init`:       initialization hook, receives the options object passed to start
+                    function; it should create and return the local state of the service;
+                    (this attr is optional)
+   - `:stop`:       termination hook, receives the service spec, internal state and
+                    allows proper resource clearing; this parameter is optional.
+   - `:receive`:    function that receives the internal state and variable number of args
+                    identifying the message sent to the service.
+   - `:timeout`:    default maximum number of milliseconds to wait for a response
+                    from the service (default 1min).
+   - `:buf-or-n`:   core.async buffer or size of buffer to use for the
+                    service communication channel (default 1024).
+  "
+  [spec]
+  (letfn [(factory [{:keys [::instances] :as options}]
```

Comment:
I've been thinking about `::instances` and I'm not sure if this is something we should care about at this level, maybe would be better to do it in a higher layer